### PR TITLE
Add support for CPLEX

### PR DESCRIPTION
**Pull request by rkersh**
With this commit a new legacy CPLEX solver has been added that
supports LP, QP, SOCP, and MIP. This is exposed via a new 1.0 style
solver under cvxpy/reductions/solvers/conic_solvers. Minor changes
have been made to the existing CPLEX solver under
cvxpy/reductions/solvers/qp_solvers for consistency. As expected, the
solver output will be hidden by default, but can be shown by setting
the verbose keyword argument to True. The warm_start solver option is
supported for linear constraints. There are two CPLEX-specific solver
options which are supported: 1) cplex_params is a keyword argument
that accepts a dictionary of parameter names (as used in the CPLEX
Python API) and parameter values; 2) cplex_filename takes a file path
which will be used to export the model to a given file format (e.g.,
SAV, LP, MPS) before solving.

For example, the following snippet will set the solver to CPLEX,
display the engine log, set a CPLEX parameter, and export the model to
LP format:

```
>>> # Set CPXPARAM_Benders_Strategy (1501) to FULL (3)
... cpx_params = {"benders.strategy": 3}
>>> prob.solve(solver=cvx.CPLEX, verbose=True,
               cplex_params=cpx_params,
               cplex_filename="example.lp")
```

### CODE REVIEW

**Comment by rkersh** [comment]
I don't think it will be a problem, but I will have to confirm with IBM legal before signing the CLA. Sorry for the delay.

**Comment by SteveDiamond** [comment]
This is fantastic! One issue that we can help you with is we aren't putting solvers in cvxpy/problems/solvers anymore. They go in cvxpy/reductions/solvers/conic_solvers. The cvxpy/problems/solvers folder is legacy code that we are working to translate to the new 1.0 style. So the new interface needs to go in cvxpy/reductions/solvers/conic_solvers and match the style there.

**Comment by rkersh** [comment]
@SteveDiamond, just to confirm, are you saying that you (or your team) will take care of the translation to the new 1.0 style? Beyond signing the CLA, is there anything else you will need from me?

**Comment by rileyjmurray** [comment]
@rkersh  I don't speak for Steven, but I don't think he'll have the time to personally convert the solver interface to the 1.0 style. 

I would normally be willing to take charge of that, but there's a backlog of deployment and infrastructure things I need to work on first. I can help guide you through the process of writing the interface in the 1.0 style, but if you don't have time to continue working on this then we can try to find an alternative way forward.

**Comment by rkersh** [comment]
@rileyjmurray: I'm happy to help, as time permits. Some guidance would be appreciated, for sure. Is there any reason why the "legacy interfaces" can't be useful in the interim?

**Comment by rkersh** [comment]
@SteveDiamond, @rileyjmurray, @bstellato, (or anyone else), I have some time to work on this. I'll see if I can get these merge conflicts fixed from above. Also, it looks like some of the new 1.0 style solvers just call the legacy solvers under the hood. For example, under `cvxpy/reductions/solvers/conic_solvers`, `xpress_conif.py:solve_via_data`, `gurobi_conif.py:solve_via_data`. That would probably be a fairly easy way to move forward here. Does that sound reasonable? Any other concerns/advice?

**Comment by SteveDiamond** [comment]
@rkersh if you don't have time to convert to the new 1.0 style interface, I guess that's ok. Obviously I would prefer if it were converted. The goal is to migrate all the old style interfaces to the 1.0 style. But we all have limited time so I understand.

**Comment by rkersh** [comment]
@bstellato, I made a few changes in `cplex_qpif.py` so that tests would pass and so the extra options are consistent with `cplex_conif.py`. I actually preferred your method of setting CPLEX parameters so I replaced what I had before with that. Do you have time for a review?

@SteveDiamond, otherwise, all checks have passed, so this pull request should be ready for general review.

File: cvxpy/tests/test_solvers.py [hunk 1]
```diff
@@ -660,3 +911,7 @@ def test_installed_solvers(self):
                 with self.assertRaises(Exception) as cm:
                     prob.solve(solver=solver)
                 self.assertEqual(str(cm.exception), "The solver %s is not installed." % solver)
+
+
+if __name__ == "__main__":
```
Comment (SteveDiamond) [inline cvxpy/tests/test_solvers.py]: What's this about?

**Comment by SteveDiamond** [comment]
Hey it's looking good! Can you explain more your changes to the unit tests?

File: cvxpy/tests/test_solvers.py (hunk 1 above)
Comment (rkersh) [reply cvxpy/tests/test_solvers.py]: That simply allows you to run `test_solvers.py` as a free standing script that accepts command line args and runs without any special test runners such as `nose`. The unittest.main method is described [here](https://docs.python.org/2/library/unittest.html#unittest.main). I was running the tests constantly while developing the CPLEX interface, so this just made my life easier, and it is harmless. I can remove those if you'd prefer.

**Comment by mickeydonkey** [comment]
Hi @rkersh  Thanks for the awesome contribution, I am playing with this PR. 
the qp test case can't pass. I notice you said there are already some fix  to made qp test case pass. Is this PR up-to-date or maybe the problem remains?

**Comment by rkersh** [comment]
@mickeydonkey, I'm not aware of any failures with commit `aeab5d2`. Can you give me more details? Which test is failing? What version of CPLEX do you have? Can you show the full traceback?

**Comment by mickeydonkey** [comment]
CPLEX I use is community edition, the version is 12.8, installed from anaconda channel ibmdecisionoptimization

`
$ conda list  |grep cplex
cplex                     12.8                     py27_0    ibmdecisionoptimization
`
I run it on my linux workstation.

here is the log, It says Q matrix is not symetric, So I guess it is some accuracy issue?

`
CPLEX Error  5012: Q is not symmetric.

Error
Traceback (most recent call last):
  File "/home/tao/miniconda2/envs/cvx/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/home/tao/Projects/cvxpy/cvxpy/tests/test_qp.py", line 101, in test_all_solvers
    self.equivalent_forms_3(solver)
  File "/home/tao/Projects/cvxpy/cvxpy/tests/test_qp.py", line 381, in equivalent_forms_3
    s = self.solve_QP(p3, solver)
  File "/home/tao/Projects/cvxpy/cvxpy/tests/test_qp.py", line 73, in solve_QP
    return problem.solve(solver=solver_name, verbose=True)
  File "/home/tao/Projects/cvxpy/cvxpy/problems/problem.py", line 246, in solve
    return solve_func(self, *args, **kwargs)
  File "/home/tao/Projects/cvxpy/cvxpy/problems/problem.py", line 360, in _solve
    kwargs)
  File "/home/tao/Projects/cvxpy/cvxpy/reductions/solvers/solving_chain.py", line 233, in solve_via_data
    solver_opts, problem._solver_cache)
  File "/home/tao/Projects/cvxpy/cvxpy/reductions/solvers/qp_solvers/cplex_qpif.py", line 152, in solve_via_data
    model.objective.set_quadratic(qmat)
  File "/home/tao/miniconda2/envs/cvx/lib/python2.7/site-packages/cplex/_internal/_subinterfaces.py", line 4118, in set_quadratic
    self.set_quadratic(_HBMatrix(args[0]))
  File "/home/tao/miniconda2/envs/cvx/lib/python2.7/site-packages/cplex/_internal/_subinterfaces.py", line 4114, in set_quadratic
    args[0].matval)
  File "/home/tao/miniconda2/envs/cvx/lib/python2.7/site-packages/cplex/_internal/_procedural.py", line 1720, in copyquad
    check_status(env, status)
  File "/home/tao/miniconda2/envs/cvx/lib/python2.7/site-packages/cplex/_internal/_procedural.py", line 303, in __call__
    raise CplexSolverError(error_string, env, status)
CplexSolverError: CPLEX Error  5012: Q is not symmetric.
`

**Comment by mickeydonkey** [comment]
By enforce the symmetry of matrix in quadratic form,I can circumvent this error.
To achieve this, the following line is modified.
https://github.com/cvxgrp/cvxpy/blob/941b1caf04dc16824e7d49fb2343eb70803c2c7e/cvxpy/atoms/matrix_frac.py#L141

Instead of pass LA.inv(P)  directly, I use 
```
        inv_p = LA.inv(P)
        return QuadForm(X, (inv_p + inv_p.T)/2)
```
It seems function LA.inv() is not tailed to symmetric matrix, and this leads to accuracy issue, which make the output inverse matrix not exactly symmetric.

**Comment by SteveDiamond** [comment]
I'm going to merge this. @mickeydonkey I'm not sure about this change. Does your script only fail for CPLEX?

**Comment by SteveDiamond** [comment]
@rkersh did you get rid of a unit test for CVXOPT in test_solvers? The merge conflicts look very strange.

**Comment by mickeydonkey** [comment]
Yes, Only this asymmetric Q error.

**Comment by rkersh** [comment]
@SteveDiamond, I have merged in the latest and resolved the conflicts you mentioned. It wasn't that I didn't have the latest `test_solvers.py`, it was the fact that new commits were pushed _after_ I rebased my work onto master.

**Comment by SteveDiamond** [comment]
Great! Thanks.

**Comment by rkersh** [comment]
@mickeydonkey, it may make sense for you to open up a new issue for problem you described above. Otherwise, it may get lost in the noise.

**Comment by rkersh** [comment]
In case anyone missed it, I answered @mickeydonkey's last question in issue #560.

**Comment by rkersh** [comment]
@rileyjmurray, @SteveDiamond: It looks like the documentation at cvxpy.org is for 1.0.6, but 1.0.8 has been released for 21 days. How does the documentation get updated on cvxpy.org website and how often? In particular, I'm asking because it would be very nice to see the documentation for the new CPLEX solver options here.
http://www.cvxpy.org/tutorial/advanced/index.html#solve-method-options

**Comment by SteveDiamond** [comment]
This is fixed. Thanks for pointing it out! I have to run a special command to update the website, which I forgot to do.

**Pull request merged**